### PR TITLE
feat: add oauth support to gsuite

### PR DESCRIPTION
**Pull request by resilience-jychp**
Allow gsuite to use oauth credentials.

Legacy method is used by default for backward compatibility.

### CODE REVIEW

**Comment by resilience-jychp** [comment]
@achantavy linter issues should be fixed now

File: docs/root/modules/gsuite/config.md [hunk 1]
```diff
@@ -15,3 +15,77 @@ Follow these steps to analyze GSuite users and groups with Cartography.
...
+
+1. Using Oauth
+
+    1. Create an App on [Google Cloud Console](https://console.cloud.google.com/)
+    1. Refer to follow documentation if needed:
+        1. https://developers.google.com/admin-sdk/directory/v1/quickstart/python
+        1. https://developers.google.com/workspace/guides/get-started
+        1. https://support.google.com/a/answer/7281227?hl=fr
+    1. Download credentials file
+    1. Use helper script for OAuth flow to obtain refresh_token
```
Comment (achantavy) [inline docs/root/modules/gsuite/config.md:27]:
```suggestion
    1. Use helper script below for OAuth flow to obtain refresh_token
```

File: docs/root/modules/gsuite/config.md [hunk 2]
```diff
@@ -15,3 +15,77 @@ Follow these steps to analyze GSuite users and groups with Cartography.
...
+    1. Serialize needed secret
+        ```
```
Comment (achantavy) [inline docs/root/modules/gsuite/config.md:29]:
```suggestion
        ```python
```

File: docs/root/modules/gsuite/config.md [hunk 3]
```diff
@@ -15,3 +15,77 @@ Follow these steps to analyze GSuite users and groups with Cartography.
...
+1. Using Oauth
+
+    1. Create an App on [Google Cloud Console](https://console.cloud.google.com/)
+    1. Refer to follow documentation if needed:
+        1. https://developers.google.com/admin-sdk/directory/v1/quickstart/python
+        1. https://developers.google.com/workspace/guides/get-started
+        1. https://support.google.com/a/answer/7281227?hl=fr
+    1. Download credentials file
+    1. Use helper script for OAuth flow to obtain refresh_token
+    1. Serialize needed secret
```
Comment (achantavy) [inline docs/root/modules/gsuite/config.md:28]: Let's either add a step here telling the user to delete the credentials.json file that was saved to disk, or add a step in the script below that deletes the file.

File: docs/root/modules/gsuite/config.md [hunk 4]
```diff
@@ -15,3 +15,77 @@ Follow these steps to analyze GSuite users and groups with Cartography.
...
+
+from google_auth_oauthlib.flow import InstalledAppFlow
+from googleapiclient.discovery import build
+
+
+scopes = ["https://www.googleapis.com/auth/admin.directory.userreadonly", "https://www.googleapis.com/auth/admin.directory.group.readonly", "https://www.googleapis.com/auth/admin.directory.group.member"]
+
+print('Go to https://console.cloud.google.com/ > API & Services > Credentials and download secrets')
+project_id = input('Provide your project ID:')
+client_id = input('Provide you Client ID:')
```
Comment (achantavy) [inline docs/root/modules/gsuite/config.md:54]:
```suggestion
client_id = input('Provide your client ID:')
```

File: docs/root/modules/gsuite/config.md [hunk 5]
```diff
@@ -15,3 +15,77 @@ Follow these steps to analyze GSuite users and groups with Cartography.
...
+client_secret = input('Provide you Client Secret:')
```
Comment (achantavy) [inline docs/root/modules/gsuite/config.md:55]:
```suggestion
client_secret = input('Provide your client secret:')
```

File: cartography/intel/gsuite/__init__.py [hunk 6]
```diff
@@ -62,23 +61,50 @@ def start_gsuite_ingestion(neo4j_session: neo4j.Session, config: Config) -> None
...
-            (
-                "Unable to initialize GSuite creds. If you don't have GSuite data or don't want to load "
-                'Gsuite data then you can ignore this message. Otherwise, the error code is: %s '
-                'Make sure your GSuite credentials are configured correctly, your credentials file (if any) is valid. '
-                'For more details see README'
-            ),
-            e,
-        )
-        return
+    if config.gsuite_auth_method == 'delegated':  #  Legacy delegated method
```
Comment (achantavy) [inline cartography/intel/gsuite/__init__.py:64]:
```suggestion
    if config.gsuite_auth_method == 'delegated':  # Legacy delegated method
        logger.info('Attempting to authenticate to GSuite using legacy delegated method')
```

File: cartography/intel/gsuite/__init__.py [hunk 7]
```diff
@@ -62,23 +61,50 @@ def start_gsuite_ingestion(neo4j_session: neo4j.Session, config: Config) -> None
...
+                    'Gsuite data then you can ignore this message. Otherwise, the error code is: %s '
+                    'Make sure your GSuite credentials file (if any) is valid. '
+                    'For more details see README'
+                ),
+                e,
+            )
+            return
+    elif config.gsuite_auth_method == 'oauth':
+        auth_tokens = json.loads(str(base64.b64decode(config.gsuite_config).decode()))
+        logger.info('SO GOOD SO FAR')
```
Comment (achantavy) [inline cartography/intel/gsuite/__init__.py:84]:
```suggestion
        logger.info('Attempting to authenticate to GSuite using OAuth')
```

File: cartography/intel/gsuite/__init__.py [hunk 8]
```diff
@@ -62,23 +61,50 @@ def start_gsuite_ingestion(neo4j_session: neo4j.Session, config: Config) -> None
...
+            credentials = GoogleCredentials(
+                None,
+                auth_tokens['client_id'],
+                auth_tokens['client_secret'],
+                auth_tokens['refresh_token'],
+                None,
+                auth_tokens['token_uri'],
+                'Cartography',
+            )
+            credentials.refresh(httplib2.Http())
```
Comment (achantavy) [inline cartography/intel/gsuite/__init__.py:95]: Just wondering, what does this do?

File: cartography/intel/gsuite/__init__.py [hunk 9]
```diff
@@ -62,23 +61,50 @@ def start_gsuite_ingestion(neo4j_session: neo4j.Session, config: Config) -> None
...
-        )
-        return
+    if config.gsuite_auth_method == 'delegated':  #  Legacy delegated method
+        try:
+            credentials = GoogleCredentials.from_stream(config.gsuite_config)
+            credentials = credentials.create_scoped(OAUTH_SCOPE)
+            credentials = credentials.create_delegated(os.environ.get('GSUITE_DELEGATED_ADMIN'))
+
+        except ApplicationDefaultCredentialsError as e:
+            logger.debug('Error occurred calling GoogleCredentials.get_application_default().', exc_info=True)
```
Comment (achantavy) [inline cartography/intel/gsuite/__init__.py:71]:
```suggestion
```

I realize this code was left over from before; we can omit it now I think.

File: cartography/intel/gsuite/__init__.py [hunk 10]
```diff
@@ -62,23 +61,50 @@ def start_gsuite_ingestion(neo4j_session: neo4j.Session, config: Config) -> None
...
+                auth_tokens['client_secret'],
+                auth_tokens['refresh_token'],
+                None,
+                auth_tokens['token_uri'],
+                'Cartography',
+            )
+            credentials.refresh(httplib2.Http())
+            credentials = credentials.create_scoped(OAUTH_SCOPE)
+        except ApplicationDefaultCredentialsError as e:
+            logger.debug('Error occurred calling GoogleCredentials.get_application_default().', exc_info=True)
```
Comment (achantavy) [inline cartography/intel/gsuite/__init__.py:98]:
```suggestion
```

Same here.

File: cartography/intel/gsuite/__init__.py [hunk 11]
```diff
@@ -62,23 +61,50 @@ def start_gsuite_ingestion(neo4j_session: neo4j.Session, config: Config) -> None
...
+            logger.error(
+                (
+                    "Unable to initialize GSuite creds. If you don't have GSuite data or don't want to load "
+                    'Gsuite data then you can ignore this message. Otherwise, the error code is: %s '
+                    'Make sure your GSuite credentials are configured correctly, your credentials are valid. '
+                    'For more details see README'
+                ),
+                e,
+            )
```
Comment (achantavy) [inline cartography/intel/gsuite/__init__.py:107]:
```suggestion
            )
            return
```

I think we want to return out of this function if this exception is hit, which will then continue on to run the other syncs.

**Review by achantavy — CHANGES_REQUESTED** [review]
Thanks for the PR! Left some nits and questions but overall looks good to me.

File: cartography/intel/gsuite/__init__.py (hunk 8 above)
Comment (resilience-jychp) [reply cartography/intel/gsuite/__init__.py:95]: We just store refresh token this step will request an access token from it.

**Review by achantavy — DISMISSED** [review]
Thank you for the PR!

**Comment by achantavy** [comment]
Just gotta remove a whitespace and remove the `on match set` and I think the CI will pass

**Review by achantavy — APPROVED** [review]
Thank you!

File: cartography/cli.py [hunk 12]
```diff
@@ -403,6 +403,23 @@ def _build_parser(self):
...
+            help=(
+                'The method used by GSuite to authenticate. delegated is the legacy one.'
+            ),
+        )
+        parser.add_argument(
+            '--gsuite-tokens-env-var',
+            type=str,
+            default='GSUITE_GOOGLE_APPLICATION_CREDENTIALS',
+            help=(
+                'The method used by GSuite to authenticate. delegated is the legacy one.'
```
Comment (achantavy) [inline cartography/cli.py:420]: Crap, I missed one thing in my review actually - what should the docstring here be? This is the same text message for `--gsuite-auth-method`.

Sorry for the back and forth.

**Review by achantavy — CHANGES_REQUESTED** [review]

**Review by achantavy — APPROVED** [review]

**Pull request merged**